pennylane/measurements/shots.py

```diff
@@ -0,0 +1,123 @@
+    ``ShotCopies``. The first value is the number of shots to execute, and the second value is the
+    number of times to repeat a circuit with that number of shots.
+
+    The ``Shots`` class exposes two properties:
+
+    * ``total_shots``, the total number of shots to be executed
+    * ``shot_vector``, the list of ``ShotCopies`` to be executed
+
+    Instances of this class are static. If an instance is passed to the constructor, that same
+    instance is returned. If an instance is constructed with a ``None`` value, ``total_shots``
+    will be ``None``, and this will suggest analytic execution.
+    """
+
+    total_shots: int = None
+    """The total number of shots to be executed."""
+
+    shot_vector: Tuple[ShotCopies] = None
+    """The list of ShotCopies to be executed. Each element is of the form (shots, copies)."""
+
+    _SHOT_ERROR = ValueError(
+        "Shots must be a single positive integer, a tuple pair of the form (shots, copies), or a sequence of these types."
+    )
+
+    _frozen = False
+
+    def __new__(cls, shots):
+        return shots if isinstance(shots, cls) else object.__new__(cls)
+
+    def __setattr__(self, name, value):
+        if self._frozen:
+            raise AttributeError(
+                "Shots is an immutable class. Consider creating a new instance if you need different shot values."
+            )
+        return super().__setattr__(name, value)
+
+    def __init__(self, shots):
+        if shots is None:
+            self.total_shots = None
+            self.shot_vector = ()
+        elif isinstance(shots, int):
+            if shots < 1:
+                raise self._SHOT_ERROR
+            self.total_shots = shots
+            self.shot_vector = (ShotCopies(shots, 1),)
+        elif isinstance(shots, Sequence):
+            if not all(valid_int(s) or valid_tuple(s) for s in shots):
+                raise self._SHOT_ERROR
+            self.__all_tuple_init__([s if isinstance(s, tuple) else (s, 1) for s in shots])
+        elif isinstance(shots, self.__class__):
+            return  # self already _is_ shots as defined by __new__
+        else:
+            raise self._SHOT_ERROR
+
+        self._frozen = True
+
+    def __all_tuple_init__(self, shots: Sequence[Tuple]):
+        res = []
+        total_shots = 0
+        current_shots, current_copies = shots[0]
+        for s in shots[1:]:
+            if s[0] == current_shots:
+                current_copies += s[1]
+            else:
+                res.append(ShotCopies(current_shots, current_copies))
+                total_shots += current_shots * current_copies
+                current_shots, current_copies = s
+        self.shot_vector = tuple(res + [ShotCopies(current_shots, current_copies)])
+        self.total_shots = total_shots + current_shots * current_copies
+
+    @property
+    def has_partitioned_shots(self):
+        """Checks if the device was instructed to perform executions with partitioned shots.
```

Comment:
I think we could also improve this documentation. This seems like a cyclical definition. "has partitioned shots means has partitioned shots". What exactly does partitioned mean in this case? Also, I think we can remove the reference to 
"device".